### DISCUSSION TITLE
UniMath User Meeting

### DISCUSSION THREAD

**Discussion by benediktahrens** [UniMath/UniMath#1635] [General]
Dear All,

I would like to propose a UniMath User Meeting, to take place online (via Zoom or similar).
The goal of this meeting is to
1. identify goals and issues with UniMath, and
2. form teams tackling the solution of these goals and issues.

Discussion of each goal should be brief and restricted to stating and delimiting the goal. Looking for solutions should be done separately.

If you would like to participate in the meeting, please fill in this doodle:

https://doodle.com/meeting/participate/id/eXQJoMoa

If you are not available at any of the time slots, please do comment here. You could mention what you are interested in and what you would like to discuss. We can then try to find a different timeslot.

Best,
Benedikt

**Comment by benediktahrens** [comment]
Newcomers are explicitly welcome and encouraged to join the UniMath User Meeting. We will try to propose projects that can serve as an entry point to formalizing mathematics in UniMath.

**Comment by benediktahrens** [comment]
The best time seems to be Friday, 5-6pm. I'll post a Zoom link here a few minutes before the start of the meeting.

**Reply by rmatthes** [reply]
Friday March 17.

**Comment by DanGrayson** [comment]
What time zone is that?

**Reply by benediktahrens** [reply]
Brussels time zone

**Comment by benediktahrens** [comment]
The Zoom URL is https://tudelft.zoom.us/j/99397354809.

**Comment by rmatthes** [comment]
[UniMath User Meeting March 2023.md](https://github.com/UniMath/UniMath/files/11003403/UniMath.User.Meeting.March.2023.md)

**Comment by rmatthes** [comment]
The above markdown file was mostly covered but at very different degrees of depth (notably with the exception of the bridges to Coq-HoTT that was skipped). I list here some remarks that came up in the discussion: The question of download from CI goes under the word "artifacts" (there might be problems of interaction with local cache, but also of size limitations for the free GitHub plan). The generic name for tools to check for stylistic errors is a "linter", and we might want to have a Coq linter here. Could there be a way to indicate to Coq which ingredients of a structure should be opaque? Could there be a way to unite definitions and lemmas that follow a specific naming pattern into a complex structure, such as a bicategory? This would be rather meta-programming for Coq. (And the other direction: can one generate templates for all the definitions and lemmas to be provided?)

This small recap does not cover the other questions (unrelated to the markdown file) and discussions of the meeting, and it does not record the expressed opinions and tasks that participants accepted.

**Reply by benediktahrens** [reply]
I think it is "artifact", not "asset".

**Reply by rmatthes** [reply]
Sorry, the right word was used in the meeting. I now change it in the message above.

**Comment by m-lindgren** [comment]
@benediktahrens FYI, here's the [overlay](https://github.com/coq/coq/blob/master/dev/ci/ci-unimath.sh) used by Coq-CI today. It seems they dropped Bicats and SubstitutionSystems 6 months ago, and they [dropped](https://github.com/coq/coq/commit/ecdd3319c78ee2947b210503b807053919ef5f94) DisplayedInserter.v way back in June 2021.

**Comment by ghost** [comment]
Regarding 
"Could there be a way to unite definitions and lemmas that follow a specific naming pattern into a complex structure, such as a bicategory? This would be rather meta-programming for Coq. (And the other direction: can one generate templates for all the definitions and lemmas to be provided?)"

When working with a notion of structured category, e.g. monoidal category, dagger category, etc, there is in general a file/folder in CategoryTheory and in Bicategories, there is the construction of the bicategory of those structured categories.
Ideally, the construction of the bicategory should not add any new information: the correct notions of functors and transformations and the laws which make them into a bicategory, should be proven already in CategoryTheory. 
Furthermore, the construction of the bicategory, which is usually done as a displayed bicategory over Cat, is of a particular form depending on e.g. the homotopy levels of the 'newly' added structure. For example, with dagger (or involutive) categories, the displayed objects form a set, the displayed morphisms form a proposition and the displayed 2-cells are contractible. Again, a proof of these homotopy-levels should already be available in CategoryTheory.
From these observations, we can see that, even though someone formalizing these structured categories might not know bicategories, it should be 'straightforward' to construct the corresponding (displayed) bicategory.
Moreover, even though there are different structures on a category, the file where the structure on the category (resp. functor, natural transformation) is defined, should be similar across the different structures, in particular the naming convention.

My proposal is to have a script that generates a template for the files on structured categories. To be more precise: One should be able to have a script/program, called something like, "CreateStructuredCat" which takes in the following arguments:
- The name of structure (lets called it Struc for now);
- The h-level of the type of objects ;
- The h-level of the type of morphisms ;
- The h-level of the type of 2-cells.
This script then creates (some of) the following files:
- CategoryTheory/Struc/Categories.v
- CategoryTheory/Struc/Functors.v
- CategoryTheory/Struc/Transformations.v
- Bicategories/DisplayedBicats/Examples/Struc.v (this file in particular uses the already existing definitions of displayed bicategories, e.g., if one would give in the h-levels prop, contr, contr, then the construction would make use of fullsub
- FunctorCategory.v, which constructs the category of structured functors and transformations, this should be completely generated and filled in from the generated statements in Functors.v and Transformations.v.
I say some of these files, because if e.g. the type of displayed 2-cells is the unit, it would not generate the Transformations file.

More (optional*) arguments could be added, such as 
- Exponentiable.v (or CartesianClosed, ...) which would construct a template if the structured functor category would form a Structured category themselves (could be extended do enriched setting, but it is already more exotic)
- (Local/Global/or both) Univalence: Which would add (admitted) lemmas that the corresponding bicat is univalent (or in the case of local univalence, adds a lemma in FunctorCategory.v that states that it is univalent.
The above 2 are optional arguments as this might not hold in general (or the user doesn't want to formalize it..)

Remark: This is a proposal, not sure if this would be worth anyone's time. It would depend, in particular, how much new structure would/could be added to Unimath (there is already quite some) and as it has a very specific use case. So if more people would think that this might be good, we should have a bunch of examples that are not yet available in UniMath.

**Reply by rmatthes** [reply]
I wonder if the language for that "script" would not rather be MetaCoq and the artefact to be invented not a plugin.

**Comment by nmvdw** [comment]
I disagree with Kobe's suggestion and I don't think that's helpful. My main disagree comes from the following two remarks.

> Ideally, the construction of the bicategory should not add any new information: the correct notions of functors and transformations and the laws which make them into a bicategory, should be proven already in CategoryTheory.

> From these observations, we can see that, even though someone formalizing these structured categories might not know bicategories, it should be 'straightforward' to construct the corresponding (displayed) bicategory.

The problem is that this presupposes one specific way of constructing the displayed bicategory, which might not be good for the purpose of univalence.

The reason is that one might want the displayed bicategory to be defined in a specific way. See the following examples

https://github.com/UniMath/UniMath/blob/master/UniMath/Bicategories/Core/Examples/StructuredCategories.v

https://github.com/UniMath/UniMath/blob/master/UniMath/Bicategories/DisplayedBicats/Examples/CwF.v

https://github.com/UniMath/UniMath/blob/master/UniMath/Bicategories/DisplayedBicats/Examples/MonadsLax.v

For each of these, a particular construction of the displayed bicategory is chosen, and the reasons are plentiful. Reusing given constructions, simplifying proofs of univalence, being able to construct displayed pseudofunctors over the identity. The same actually also happens for monoidal categories: https://github.com/UniMath/UniMath/tree/master/UniMath/Bicategories/MonoidalCategories/UnivalenceMonCat

Automating this does not make sense, because that eliminates the nuances in how to construct the displayed bicategory. For monoidal categories, one might even consider making it an instance of a more general construction, namely pseudomonoids in a bicategory (note that a similar idea is used in https://github.com/UniMath/UniMath/blob/master/UniMath/Bicategories/DisplayedBicats/Examples/MonadsLax.v where monads are defined for arbitrary bicategories and not just for categories).

In addition, there could be reasons to put certain kinds of categegorical structures in the same file, as done in https://github.com/UniMath/UniMath/blob/master/UniMath/Bicategories/Core/Examples/StructuredCategories.v

Furthermore, many complicated categorical structures will be built from combining other categorical structures. However, automatically generating them in the suggested way, will not do that.

I think the usability of the suggested script will turn out to be very limited and that ultimately, it will not be used. In addition, it might even generate files/constructions that are undesired: if one uses this for symmetric monoidal categories, then I think that the resulting files will not be usable.
